app/apps/server/communication/rest.js

```diff
@@ -92,15 +98,23 @@ export class AppsRestApi {
 					return API.v1.success(result.data);
 				}
 
-				if (this.queryParams.buildBuyUrl && this.queryParams.appId) {
+				if (this.queryParams.buildExternalUrl && this.queryParams.appId) {
 					const workspaceId = settings.get('Cloud_Workspace_Id');
 
+					if (!this.queryParams.purchaseType || !purchaseTypes.has(this.queryParams.purchaseType)) {
+						return API.v1.failure({ error: 'Invalid purchase type' });
+					}
+
 					const token = getUserCloudAccessToken(this.getLoggedInUser()._id, true, 'marketplace:purchase', false);
 					if (!token) {
 						return API.v1.failure({ error: 'Unauthorized' });
 					}
 
-					return API.v1.success({ url: `${ baseUrl }/apps/${ this.queryParams.appId }/buy?workspaceId=${ workspaceId }&token=${ token }` });
+					return API.v1.success({
+						url: `${ baseUrl }/apps/${ this.queryParams.appId }/${
+							this.queryParams.purchaseType === 'buy' ? this.queryParams.purchaseType : 'subscribe'
```

Comment:
Why need this check since it's already requiring the property to have only valid values on line 104?

Reply:
The `purchaseType` property from the app does _not_ map directly to the endpoint's path, so this is not exaclty a validation, more of a mapping